### PR TITLE
feat/TV-34-implement-update-user-permission-usecase

### PR DESCRIPTION
**Pull request by Arvicss**
## PR Question
**Do you have specific questions or issues you would like to address with the reviewer?**
- Do I need to test [userRouter](https://github.com/ThesisVault/thesis-vault/blob/feat/TV-34-implement-update-user-permission-usecase/src/modules/user/src/infrastructure/http/routes/userRouter.ts)? If yes, how?

## Description
- implement updateUserPermissionUseCase
- implement updateUserPermissionController
- define application errors
- create test cases

## PR/Code Author Checklist
**Note:** please write "x" within the brackets if item applies e.g. [x]

- [x] I set the PR title based on the following format: [{task_type}] {task_id}: {task_name} . e.g. [feat] TV-X: add-new-functionality (camelCase)

- [x] I set the branch name based on the following format: {branch_type}/{task_id}-{task_name}. e.g feat/LM-X-add-new-functionality

- [x] I set KuroXI as the reviewer of this PR

- [x] I set myself as an assignees of this PR

- [x] I set the correct labels of this PR

- [x] All new and existing tests passed

- [x] I did not skip/commented-out/removed any existing tests

- [x] I have performed a self-review of my code

### CODE REVIEW

File: src/modules/user/tests/domain/models/user/classes/user.test.ts [hunk 1]
```diff
@@ -5,6 +5,7 @@ import { UserRole } from "@/modules/user/src/domain/models/user/classes/userRole
 import { Permissions } from "@/modules/user/src/domain/models/user/shared/permission/permissions";
 import { Roles } from "@/modules/user/src/domain/models/user/shared/permission/roles";
 import { faker } from "@faker-js/faker";
+import { createUserDomainObject } from "@/modules/user/tests/utils/createUserDomainObject";
```
Comment (KuroXI) [inline src/modules/user/tests/domain/models/user/classes/user.test.ts:8]: Please organize your imports. [Optimize Imports in WebStorm](https://blog.jetbrains.com/webstorm/2018/05/optimize-imports-in-webstorm/)

File: src/modules/user/tests/domain/models/user/service/userPermissionService.test.ts [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+import { Permissions } from "@/modules/user/src/domain/models/user/shared/permission/permissions";
+
+describe("UserPermissionService", () => {
+  let userPermissionService: UserPermissionService;
+  
+  beforeAll(async () => {
+    userPermissionService = new UserPermissionService();
+  });
+  
+  afterEach(async () => {
```
Comment (KuroXI) [inline src/modules/user/tests/domain/models/user/service/userPermissionService.test.ts:13]: I think `deleteMany` is not necessary here

File: src/modules/user/tests/infrastructure/http/controllers/updateUserPermissionController.test.ts [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+describe('UpdateUserPermissionController', () => {
+  let controller: UpdateUserPermissionController;
+  let userPermissionServiceMock: jest.Mocked<UserPermissionService>;
+  let updateUserPermissionUseCaseMock: jest.Mocked<UpdateUserPermissionUseCase>;
+  
+  beforeEach(() => {
+    userPermissionServiceMock = new UserPermissionService() as jest.Mocked<UserPermissionService>;
+    updateUserPermissionUseCaseMock = new UpdateUserPermissionUseCase() as jest.Mocked<UpdateUserPermissionUseCase>;
+    
+    controller = new UpdateUserPermissionController(userPermissionServiceMock, updateUserPermissionUseCaseMock);
```
Comment (KuroXI) [inline src/modules/user/tests/infrastructure/http/controllers/updateUserPermissionController.test.ts:22]: why initialize the service and useCase here when it's already been initialized inside controller?

File: src/modules/user/tests/repositories/updateUser.test.ts [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+import { createUserDomainObject } from "@/modules/user/tests/utils/createUserDomainObject";
+
+describe("UserRepository.updateUser", () => {
+  let userRepository: UserRepository;
+  
+  beforeAll(async () => {
+    userRepository = new UserRepository();
+  });
+  
+  afterEach(async () => {
```
Comment (KuroXI) [inline src/modules/user/tests/repositories/updateUser.test.ts:14]: I think deleteMany is not necessary here

File: src/modules/user/tests/repositories/updateUsers.test.ts [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+import { createUserDomainObject } from "@/modules/user/tests/utils/createUserDomainObject";
+
+describe("UserRepository.updateUsers", () => {
+  let userRepository: UserRepository;
+  
+  beforeAll(async () => {
+    userRepository = new UserRepository();
+  });
+  
+  afterEach(async () => {
```
Comment (KuroXI) [inline src/modules/user/tests/repositories/updateUsers.test.ts:14]: I think deleteMany is not necessary here

File: src/modules/user/tests/useCases/updateUserPermissionUseCase.test.ts [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+
+	beforeEach(() => {
+		userRepositoryMock = {
+			getUserById: jest.fn(),
+			getUsersByIds: jest.fn(),
+			updateUser: jest.fn(),
+			updateUsers: jest.fn(),
+		} as jest.Mocked<IUserRepository>;
+
+		updateUserPermissionUseCase = new UpdateUserPermissionUseCase(userRepositoryMock);
```
Comment (KuroXI) [inline src/modules/user/tests/useCases/updateUserPermissionUseCase.test.ts:21]: Why mock the repository?

**Review by KuroXI — CHANGES_REQUESTED** [review]
Please also add a test for the endpoint you created.
```
- src
- tests
  - api
    - user
      - updateUserPermission.test.ts
```

**Comment by KuroXI** [comment]
@Arvicss To test your userRouter, we can import `auth` and `trpc` in our test cases since they're on the shared folder.

- `src/shared/infrastructure/authentication/index.ts`
- `src/shared/infrastructure/trpc/api/server.ts`.

Then we can call the `getServerAuthSession()` from auth and `updateUserPermission` from userRouter.

We also need to create a session in `beforeAll(() => {})` so that we can use the protectedProcedure you created.

Add a test also for the unauthorized request and user that does not have the required permission.

```ts
describe("userRouter updateUserPermission endpoint", () => {
  beforeAll(() => {
    // authenticate user
  })

  describe("User is authorized", () => {
    it("should return an error if user does not have MANAGE_USER permission")
    // ...
  })
  
  describe("User is un-authorized", () => {
    it("should return an error if user is not authenticated")
    // ...
  })
})
```

This is solely an idea, feel free to improve this.

File: src/modules/user/src/repositories/userRepository.ts [hunk 7]
```diff
@@ -47,4 +47,33 @@ export class UserRepository implements IUserRepository {
 			isDeleted: false,
 		};
 	}
+
+	async updateUser(data: IUser): Promise<IUser | null> {
```
Comment (KuroXI) [inline src/modules/user/src/repositories/userRepository.ts:51]: Please add this to `IUserRepository` and always make sure private method are in the bottom

File: src/modules/user/src/domain/models/user/service/userPermissionService.ts
Comment (KuroXI) [inline src/modules/user/src/domain/models/user/service/userPermissionService.ts:1]: You can delete this file and use `src/modules/user/src/domain/services/userPermissionService.ts`

File: src/modules/user/src/shared/permissions.ts [hunk 8]
```diff
@@ -20,5 +20,4 @@ export const Permissions = {
 	// ! UPDATE THIS WHENEVER YOU ADD NEW PERMISSION
 	ALL: (1 << 12) - 1,
 } as const;
-
-export type PermissionKeys = keyof typeof Permissions;
+export type PermissionKeys =  keyof typeof Permissions;
```
Comment (KuroXI) [inline src/modules/user/src/shared/permissions.ts:23]: Fix your formatter

File: src/modules/user/tests/domain/models/user/service/userPermissionService.test.ts
Comment (KuroXI) [inline src/modules/user/tests/domain/models/user/service/userPermissionService.test.ts:1]: You can now delete this file and please rebase to the master

File: src/modules/user/tests/utils/createUserDomainObject.ts
Comment (KuroXI) [inline src/modules/user/tests/utils/createUserDomainObject.ts:1]: You can now remove this file

File: src/modules/user/src/useCases/updateUserPermissionUseCase.ts
Comment (KuroXI) [inline src/modules/user/src/useCases/updateUserPermissionUseCase.ts:1]: User permission now have 2 properties which is `allowPermissions` and `denyPermissions`.

Please update your implementation and sync the sequence diagram to your implementation.

**Review by KuroXI — CHANGES_REQUESTED** [review]
Please rebase into master

**Comment by Arvicss** [comment]
I'm having issues getting the tRPC route test to run. I've tried changing test configuration multiple times, but  still encountering the same error.


FAIL   node-tests  src/modules/user/tests/api/user/updatePermission.test.ts
  ● Test suite failed to run

    Jest encountered an unexpected token

    Jest failed to parse a file. This happens e.g. when your code or its dependencies use non-standard JavaScript syntax, or when Jest is not configured to support such syntax.

    Out of the box Jest supports Babel, which will be used to transform your files into valid JS based on your Babel configuration.

    By default "node_modules" folder is ignored by transformers.

    Here's what you can do:
     • If you are trying to use ECMAScript Modules, see https://jestjs.io/docs/ecmascript-modules for how to enable it.
     • If you are trying to use TypeScript, see https://jestjs.io/docs/getting-started#using-typescript
     • To have some of your "node_modules" files transformed, you can specify a custom "transformIgnorePatterns" in your config.
     • If you need a custom transformation specify a "transform" option in your config.
     • If you simply want to mock your non-JS modules (e.g. binary assets) you can stub them out with the "moduleNameMapper" config option.

    You'll find more details and examples of these config options in the docs:
    https://jestjs.io/docs/configuration
    For information about custom transformations, see:
    https://jestjs.io/docs/code-transformation

    Details:

    F:\Programming Projects\Web\thesis-vault\node_modules\.pnpm\superjson@2.2.1\node_modules\superjson\dist\index.js:1
    ({"Object.<anonymous>":function(module,exports,require,__dirname,__filename,jest){import { ClassRegistry } from './class-registry.js';
                                                                                      ^^^^^^

    SyntaxError: Cannot use import statement outside a module

      1 | import { TRPCError, initTRPC } from "@trpc/server";
    > 2 | import superjson from "superjson";
        | ^
      3 | import { ZodError } from "zod";
      4 | import { getServerAuthSession } from "../authentication";
      5 | import { db } from "../database";

      at Runtime.createScriptFromCode (node_modules/.pnpm/jest-runtime@29.7.0/node_modules/jest-runtime/build/index.js:1505:14)
      at Object.<anonymous> (src/shared/infrastructure/trpc/index.ts:2:1)
      at Object.<anonymous> (src/shared/infrastructure/trpc/routes.ts:1:1)
      at Object.<anonymous> (src/modules/user/tests/api/user/updatePermission.test.ts:3:1)

File: src/modules/user/tests/useCases/updateUserPermissionUseCase.test.ts [hunk 9]
```diff
@@ -0,0 +1,76 @@
...
+			await updateUserPermissionUseCase.execute(request);
+		} catch (error) {
+			errorMessage = (error as Error).message;
+			expect(error).toBeInstanceOf(BadRequestError);
+		}
+		
+		expect(errorMessage).toContain("Failed to update permission");
+	});
+	
+	it("should successfully update user permission and return userId", async () => {
```
Comment (KuroXI) [inline src/modules/user/tests/useCases/updateUserPermissionUseCase.test.ts:55]: Why are we not checking if the `denyPermission` got updated? Please double check your implementation

**Review by KuroXI — CHANGES_REQUESTED** [review]

File: src/modules/user/src/useCases/updateUserPermissionUseCase.ts [hunk 10]
```diff
@@ -0,0 +1,45 @@
+import { UserPermission } from "@/modules/user/src/domain/models/user/classes/userPermission";
+import type { UpdateUserPermissionDTO } from "@/modules/user/src/dtos/userDTO";
+import {
+	type IUserRepository,
+	UserRepository,
+} from "@/modules/user/src/repositories/userRepository";
+import { BadRequestError, NotFoundError, UnexpectedError } from "@/shared/core/errors";
+
+export class UpdateUserPermissionUseCase {
```
[the remark this answers is not in the document]
Comment (Arvicss) [reply src/modules/user/src/useCases/updateUserPermissionUseCase.ts:9]: image

File: src/modules/user/tests/repositories/user/updateUser.test.ts [hunk 11]
```diff
@@ -0,0 +1,46 @@
+import { UserRepository } from "@/modules/user/src/repositories/userRepository";
+import { Permissions } from "@/modules/user/src/shared/permissions";
+import { db } from "@/shared/infrastructure/database";
+import { seedUser } from "@/modules/user/tests/utils/user/seedUser";
+import { UserMapper } from "@/modules/user/src/mappers/userMapper";
+import { createUserDomainObject } from "@/modules/user/tests/utils/user/createUserDomainObject";
+import { faker } from "@faker-js/faker";
+
+describe("UserRepository.updateUser", () => {
+  let userRepository: UserRepository;
```
Comment (KuroXI) [inline src/modules/user/tests/repositories/user/updateUser.test.ts:10]: use `IUserRepository`

File: src/modules/user/tests/repositories/user/updateUsers.test.ts [hunk 12]
```diff
@@ -0,0 +1,62 @@
+import { Permissions } from "@/modules/user/src/shared/permissions";
+import { db } from "@/shared/infrastructure/database";
+import { seedUser } from "@/modules/user/tests/utils/user/seedUser";
+import { UserMapper } from "@/modules/user/src/mappers/userMapper";
+import { UserRepository } from "@/modules/user/src/repositories/userRepository";
+import { createUserDomainObject } from "@/modules/user/tests/utils/user/createUserDomainObject";
+import { faker } from "@faker-js/faker";
+
+describe("UserRepository.updateUsers", () => {
+	let userRepository: UserRepository;
```
Comment (KuroXI) [inline src/modules/user/tests/repositories/user/updateUsers.test.ts:10]: use `IUserRepository`

File: src/modules/user/tests/useCases/updateUserPermissionUseCase.test.ts [hunk 13]
```diff
@@ -0,0 +1,77 @@
+import type { UpdateUserPermissionDTO } from "@/modules/user/src/dtos/userDTO";
+import { UserRepository } from "@/modules/user/src/repositories/userRepository";
+import { Permissions } from "@/modules/user/src/shared/permissions";
+import { UpdateUserPermissionUseCase } from "@/modules/user/src/useCases/updateUserPermissionUseCase";
+import { seedUser } from "@/modules/user/tests/utils/user/seedUser";
+import { BadRequestError, NotFoundError } from "@/shared/core/errors";
+import { faker } from "@faker-js/faker";
+
+describe("UpdateUserPermissionUseCase", () => {
+	const updateUserPermissionUseCase = new UpdateUserPermissionUseCase();
```
Comment (KuroXI) [inline src/modules/user/tests/useCases/updateUserPermissionUseCase.test.ts:10]: Be consistent on how you initialize a class in test

Use reference: `src/modules/user/tests/repositories/user/updateUsers.test.ts`

File: src/modules/user/tests/infrastructure/http/controllers/updateUserPermissionController.test.ts [hunk 14]
```diff
@@ -0,0 +1,45 @@
+import type { UpdateUserPermissionDTO } from "@/modules/user/src/dtos/userDTO";
+import {
+  UpdateUserPermissionController
+} from "@/modules/user/src/infrastructure/http/controllers/updateUserPermissionController";
+import { Permissions } from "@/modules/user/src/shared/permissions";
+import { seedUser } from "@/modules/user/tests/utils/user/seedUser";
+import { faker } from "@faker-js/faker";
+
+describe("updateUserPermissionController", () => {
+  const updateUserPermissionController = new UpdateUserPermissionController();
```
Comment (KuroXI) [inline src/modules/user/tests/infrastructure/http/controllers/updateUserPermissionController.test.ts:10]: Be consistent on how you initialize a class in test

Use reference: `src/modules/user/tests/repositories/user/updateUsers.test.ts`

**Review by KuroXI — CHANGES_REQUESTED** [review]

File: src/modules/user/src/infrastructure/http/controllers/updateUserPermissionController.ts [hunk 15]
```diff
@@ -0,0 +1,29 @@
+import { UserPermissionService } from "@/modules/user/src/domain/services/userPermissionService";
+import type { UpdateUserPermissionDTO } from "@/modules/user/src/dtos/userDTO";
+import { UpdateUserPermissionUseCase } from "@/modules/user/src/useCases/updateUserPermissionUseCase";
+import { UnauthorizedError } from "@/shared/core/errors";
+import { BaseController } from "@/shared/infrastructure/trpc/models/baseController";
+
+export class UpdateUserPermissionController extends BaseController<UpdateUserPermissionDTO, string> {
+  private userPermissionService: UserPermissionService;
```
Comment (KuroXI) [inline src/modules/user/src/infrastructure/http/controllers/updateUserPermissionController.ts:8]: Use the interface of the class not the class itself.

e.g `IUserPermissionService`

**Review by KuroXI — CHANGES_REQUESTED** [review]

**Review by KuroXI — APPROVED** [review]

**Pull request merged**